### PR TITLE
8321248: ClassFile API ClassModel::verify is inconsistent with the rest of the API

### PR DESCRIPTION
**Pull request by asotona**
ClassFile API `ClassModel::verify` accepts `ClassHierarchyResolver` as an optional argument and does not respect `ClassFile.ClassHierarchyResolverOption` of the actual context.
Parsing, building and transforming take options from the actual `ClassFile` context and verification should follow the same API pattern.

This patch removes `ClassModel::verify` methods and introduces new top level methods:
```
List<VerifyError> ClassFile::verify(ClassModel model);
List<VerifyError> ClassFile::verify(byte[] bytes);
List<VerifyError> ClassFile::verify(Path path);
```

Impact of the API change is minimal as the API has not been released yet.

Please review.

Thanks,
Adam

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8321327](https://bugs.openjdk.org/browse/JDK-8321327) to be approved

### Issues
 * [JDK-8321248](https://bugs.openjdk.org/browse/JDK-8321248): ClassFile API ClassModel::verify is inconsistent with the rest of the API (**Bug** - P3)
 * [JDK-8321327](https://bugs.openjdk.org/browse/JDK-8321327): ClassFile API ClassModel::verify is inconsistent with the rest of the API (**CSR**)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16947/head:pull/16947` \
`$ git checkout pull/16947`

Update a local copy of the PR: \
`$ git checkout pull/16947` \
`$ git pull https://git.openjdk.org/jdk.git pull/16947/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16947`

View PR using the GUI difftool: \
`$ git pr show -t 16947`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16947.diff">https://git.openjdk.org/jdk/pull/16947.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16947#issuecomment-1838424566)

### CODE REVIEW

File: test/jdk/jdk/classfile/VerifierSelfTest.java [hunk 1]
```diff
@@ -51,41 +47,10 @@ void testVerify() throws IOException {
...
-                            ClassFile.of().parse(path).verify(null);
+                            ClassFile.of().verify(path);
                         } catch (IOException e) {
                             throw new AssertionError(e);
                         }
                     });
     }
-
-    @Test
-    void testFailedDump() throws IOException {
```
Comment (mcimadamore) [inline test/jdk/jdk/classfile/VerifierSelfTest.java:62]: Why is this removed?

Comment (asotona) [reply test/jdk/jdk/classfile/VerifierSelfTest.java:62]: Dump (print) of the classfile to an optional log (Consumer<String> argument) has been removed from the API.
It was a relic from early phase of the ClassFile API development and it has no use except for this test.
This functionality can be replaced by explicit use of ClassPrinter.

Comment (asotona) [reply test/jdk/jdk/classfile/VerifierSelfTest.java:62]: I'll rename the test to "testFailed" and keep it without the "dump" park.

**Review by lahodaj — APPROVED** [review]
I guess the tests could be made much more tight. E.g. by not only checking passes/fails verification, but also checking the exact errors reported (for tests that just check that verify "failed", without checking the exact errors, it may happen the verification fails for a wrong reason). And the tests could probably be made much broader.

But I don't think that needs to happen here.

**Review by mcimadamore — APPROVED** [review]

**Comment by asotona** [comment]
Thank you for the reviews.

Yes, next step is to improve verifier range and also cover it with negative tests (see https://github.com/openjdk/jdk/pull/16809/files#diff-a23859572e86e946a9ce66361e64e4b7e4473f134bb49a248fccd70d2ac96ea2)
